tests/test_cusparseSolver.cpp

```diff
@@ -0,0 +1,132 @@
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <config.h>
+
+#define BOOST_TEST_MODULE OPM_test_cusparseSolver
+#include <boost/test/unit_test.hpp>
+#include <boost/version.hpp>
+
+#include <dune/common/version.hh>
+
+#if DUNE_VERSION_NEWER(DUNE_ISTL, 2, 6) && \
+    BOOST_VERSION / 100 % 1000 > 48
+
+#include <opm/simulators/linalg/bda/BdaBridge.hpp>
+
+#include <dune/common/fvector.hh>
+#include <dune/istl/bvector.hh>
+#include <dune/istl/bcrsmatrix.hh>
+#include <dune/istl/matrixmarket.hh>
+#include <dune/istl/solver.hh>
+
+#include <boost/property_tree/json_parser.hpp>
+#include <boost/property_tree/ptree.hpp>
+
+
+template <int bz>
+Dune::BlockVector<Dune::FieldVector<double, bz>>
+testCusparseSolver(const boost::property_tree::ptree& prm, const std::string& matrix_filename, const std::string& rhs_filename)
+{
+    using Matrix = Dune::BCRSMatrix<Dune::FieldMatrix<double, bz, bz>>;
+    using Vector = Dune::BlockVector<Dune::FieldVector<double, bz>>;
+    Matrix matrix;
+    {
+        std::ifstream mfile(matrix_filename);
+        if (!mfile) {
+            throw std::runtime_error("Could not read matrix file");
+        }
+        readMatrixMarket(matrix, mfile);
+    }
+    Vector rhs;
+    {
+        std::ifstream rhsfile(rhs_filename);
+        if (!rhsfile) {
+            throw std::runtime_error("Could not read rhs file");
+        }
+        readMatrixMarket(rhs, rhsfile);
+    }
+
+    const int linear_solver_verbosity = prm.get<int>("verbosity");
+    const int maxit = prm.get<int>("maxiter");
+    const double tolerance = prm.get<double>("tol");
+    const std::string opencl_ilu_reorder("none"); // unused
+    const int platformID = 0;                     // unused
+    const int deviceID = 0;
+    const std::string gpu_mode("cusparse");
+    const std::string fpga_bitstream("empty");    // unused
+    Dune::InverseOperatorResult result;
+
+    Vector x(rhs.size());
+    Opm::WellContributions wellContribs("cusparse");
+    try {
+        Opm::BdaBridge<Matrix, Vector, bz> bridge(gpu_mode, fpga_bitstream, linear_solver_verbosity, maxit, tolerance, platformID, deviceID, opencl_ilu_reorder);
+
+        bridge.solve_system(&matrix, rhs, wellContribs, result);
+        bridge.get_result(x);
+    } catch (const std::logic_error& error) {
+        BOOST_WARN_MESSAGE(true, error.what());
+    }
```

Comment:
I think this is the wrong place to catch the exception or at least you need to rethrow it in the catch clause. Otherwise or now we return an invalid solution and the caller cannot notice that there was a problem

Reply:
But otherwise the test fails.
The test is meant to test the solver, if the solver cannot run, because things are not installed, should it fail?
Perhaps CMake should not add the test if there is no device.

Reply:
CMake is probably not the correct place. You might compile on a system without devices but OpenCL and Cusparse installed but run on one that has devices.